core/shared/src/main/scala/zio/FiberRefs.scala

```diff
@@ -109,130 +109,175 @@ final class FiberRefs private (
-                val newEntry = parentStack match {
-                  case StackEntry(parentFiberId, _, parentVersion) :: tail =>
-                    if (parentFiberId == fiberId)
-                      Value(::(StackEntry(parentFiberId, newValue, parentVersion + 1), tail), parentDepth)
-                    else
-                      Value(::(StackEntry(fiberId, newValue, 0), parentStack), parentDepth + 1)
-                }
-                parentFiberRefs.updated(ref, newEntry)
-              }
+    val newFiberRefLocals = childFiberRefs.foldLeft(parentFiberRefs) { case (parentFiberRefs, (ref, childStack)) =>
+      if (childStack.fiberId == fiberId) parentFiberRefs
+      else {
+        val childFiberRef          = ref.asInstanceOf[FiberRef[Any]]
+        val childFiberInitialValue = childFiberRef.initial
+        val childStackCurrentValue = childStack.value
+
+        parentFiberRefs.getOrElse(childFiberRef, null) match {
+          case null =>
+            val emptyChildStack = childStackCurrentValue == childFiberInitialValue
+            if (emptyChildStack) parentFiberRefs
+            else
+              parentFiberRefs.updated(
+                childFiberRef,
+                FiberRefStack.init(fiberId, childFiberRef.join(childFiberInitialValue, childStackCurrentValue))
+              )
+          case parentStack =>
+            val ancestor =
+              findAncestor(childFiberInitialValue)(
+                parentStack.stack(),
+                parentStack.depth,
+                childStack.stack(),
+                childStack.depth
+              )
+            val patch    = childFiberRef.diff(ancestor, childStackCurrentValue)
+            val oldValue = parentStack.value
+            val newValue = childFiberRef.join(oldValue, childFiberRef.patch(patch)(oldValue))
+
+            if (oldValue == newValue) parentFiberRefs
+            else {
+              val newStack =
+                if (parentStack.fiberId == fiberId) parentStack.updateValue(newValue)
+                else parentStack.put(fiberId, newValue)
+
+              parentFiberRefs.updated(childFiberRef, newStack)
             }
         }
+      }
     }
 
-    if (self.fiberRefLocals eq fiberRefLocals0) self
-    else FiberRefs(fiberRefLocals0)
+    if (self.fiberRefLocals eq newFiberRefLocals) self
+    else FiberRefs(newFiberRefLocals)
   }
 
   def setAll(implicit trace: Trace): UIO[Unit] =
     ZIO.foreachDiscard(fiberRefs) { fiberRef =>
       fiberRef.asInstanceOf[FiberRef[Any]].set(getOrDefault(fiberRef))
     }
 
-  override final def toString(): String = fiberRefLocals.mkString("FiberRefLocals(", ",", ")")
+  override def toString: String = fiberRefLocals.mkString("FiberRefLocals(", ",", ")")
 
   def updatedAs[@specialized(SpecializeInt) A](
     fiberId: FiberId.Runtime
   )(fiberRef: FiberRef[A], value: A): FiberRefs = {
-    val oldEntry = fiberRefLocals.getOrElse(fiberRef, null)
+    val oldStack = fiberRefLocals.getOrElse(fiberRef, null)
```

Comment:
You will need to cast this, otherwise the inner type will be inferred as Any and it'll cause boxing of the Int in the specialized function

```suggestion
    val oldStack = fiberRefLocals.getOrElse(fiberRef, null).asInstanceOf[FiberRefStack[A]]
```